include/s3select_functions.h

```diff
@@ -643,226 +774,305 @@ struct _fn_to_timestamp : public base_function
-    int args_size = args->size();
+    param_validation(args);
 
-    if (args_size < 3)
-    {
-      throw base_s3select_exception("add_to_timestamp should have 3 parameters");
-    }
+    new_ptime = *val_ts.timestamp();
+    new_ptime += boost::gregorian::months( val_quantity.i64() );
+    result->set_value( &new_ptime );
+    return true;
+  }
+};
 
-    base_statement* date_part = *iter;
-    val_date_part = date_part->eval();//TODO could be done once?
+struct _fn_add_day_to_timestamp : public base_date_add
+{
+  boost::posix_time::ptime new_ptime;
 
-    if(val_date_part.is_string()== false)
-    {
-      throw base_s3select_exception("first parameter should be string");
-    }
+  bool operator()(bs_stmt_vec_t* args, variable* result) override
+  {
+    param_validation(args);
 
-    iter++;
-    base_statement* quan = *iter;
-    val_quantity = quan->eval();
+    new_ptime = *val_ts.timestamp();
+    new_ptime += boost::gregorian::days( val_quantity.i64() );
+    result->set_value( &new_ptime );
+    return true;
+  }
+};
 
-    if (val_quantity.is_number() == false)
-    {
-      throw base_s3select_exception("second parameter should be number");  //TODO what about double?
-    }
+struct _fn_add_hour_to_timestamp : public base_date_add
+{
+  boost::posix_time::ptime new_ptime;
 
-    iter++;
-    base_statement* ts = *iter;
-    val_timestamp = ts->eval();
+  bool operator()(bs_stmt_vec_t* args, variable* result) override
+  {
+    param_validation(args);
 
-    if(val_timestamp.is_timestamp() == false)
-    {
-      throw base_s3select_exception("third parameter should be time-stamp");
-    }
+    new_ptime = *val_ts.timestamp();
+    new_ptime += boost::posix_time::hours( val_quantity.i64() );
+    result->set_value( &new_ptime );
+    return true;
+  }
+};
 
-    new_ptime = *val_timestamp.timestamp();
+struct _fn_add_minute_to_timestamp : public base_date_add
+{
+  boost::posix_time::ptime new_ptime;
 
-    if( strcmp(val_date_part.str(), "year")==0 )
-    {
-      new_ptime += boost::gregorian::years( val_quantity.i64() );
-      result->set_value( &new_ptime );
-    }
-    else if( strcmp(val_date_part.str(), "month")==0 )
-    {
-      new_ptime += boost::gregorian::months( val_quantity.i64() );
-      result->set_value( &new_ptime );
-    }
-    else if( strcmp(val_date_part.str(), "day")==0 )
-    {
-      new_ptime += boost::gregorian::days( val_quantity.i64() );
-      result->set_value( &new_ptime );
-    }
-    else
-    {
-      throw base_s3select_exception( std::string(val_date_part.str() + std::string(" is not supported for add")).c_str());
-    }
+  bool operator()(bs_stmt_vec_t* args, variable* result) override
+  {
+    param_validation(args);
 
+    new_ptime = *val_ts.timestamp();
+    new_ptime += boost::posix_time::minutes( val_quantity.i64() );
+    result->set_value( &new_ptime );
     return true;
   }
+};
+
+struct _fn_add_second_to_timestamp : public base_date_add
+{
+  boost::posix_time::ptime new_ptime;
```

Comment:
/boost::posix_time::ptime new_ptime;/ 
and new_ptime = *val_ts.timestamp(); could be a part of parent class